Limit declaration of C primitives in runtime headers

dra27 opened the pull request:
Another piece of tidying (very nearly at the stage that `CAMLexport` and `CAMLprim` match up with headers in a predictable fashion!).

It turns out that there are very few cases where C primitives actually need to be called directly from C, and so declared in headers. This PR contains two tidying commits:

- Makes `Raw_spacetime_lib` directly call `caml_input_value_to_outside_heap` rather than declaring the single-line C stub `caml_spacetime_unmarshal_trie` which simply calls that C primitive.
- Removes declarations for primitives (all inside `CAML_INTERNALS`) which are not actually referenced anywhere (I did a GitHub code search on these as well)

and two refactoring commits:

- Rather than C code calling `caml_sys_exit` having encoded the exit status, the actual functionality of `caml_sys_exit` is moved to `caml_do_exit` which takes an `int` (I resisted the temptation to call the function `caml_start_exit` - it could also be called `caml_exit`?).
- Slightly buggily, the manual outlines calling the primitive `caml_record_backtrace` from C but doesn't actually explain how to declare it (it's internal in `backtrace.h`). As it happens, from C you're pretty much only ever going to want to turn backtraces *on* so I added `caml_record_backtraces` which is declared publicly in `backtrace.h` and then is used elsewhere in the runtime (since the runtime also only ever turns them on).

At the end of this process, we have exactly four primitives which actually have to be exported for use from C:
- `caml_ml_open_descriptor_in` and `caml_ml_open_descriptor_out` (the Unix library needs them)
- `caml_get_public_method` (that's used in C as well as in the Object system's runtime)
- `caml_set_oo_id` (which is used in `intern.c`)

It might be a nice piece of future work to generate `caml/primitives.h` in the same way as `prims.c`, but it's a non-trivial and not strictly necessary at the moment...

dra27 committed:
Remove unnecessary declarations of primitives

These are inside CAML_INTERNALS blocks

dra27 committed:
Introduce caml_do_exit

For calling from C, eliminating the need to export caml_sys_exit

xavierleroy commented on manual/manual/cmds/intf-c.etex:
```diff
@@ -1712,7 +1712,8 @@ information is available, but the backtrace mechanism needs to be
 one of the OCaml modules.  This can also be achieved from the C side
-by calling "caml_record_backtrace(Val_int(1));" in the OCaml-C glue code.
+by calling "caml_record_backtraces();" in the OCaml-C glue code.
+("ocaml_receord_backtraces" is declared in "backtrace.h")
```
Typo "receord"

dra27 replied:
Yuck, how many typos can one fit in one added sentence :-/ Fixed (and the `ocaml_` bit too!)

xavierleroy commented on runtime/caml/backtrace.h:
```diff
@@ -52,7 +62,8 @@
  /* [Caml_state->backtrace_active] is non zero iff backtraces are recorded.
- * This variable must be changed with [caml_record_backtrace].
+ * This variable must be changed with [caml_record_backtrace] in OCaml or
+ * caml_record_backtraces in C.
```
For consistency: `[caml_record_backtraces]` ? or even `[caml_record_backtraces()]` ?

dra27 replied:
Done

xavierleroy reviewed (commented):
This looks good to me, assuming no third-party C code ever calls `caml_sys_exit`.  I trust you grepped for this?

dra27 committed:
Introduce caml_record_backtraces

The manual suggests calling caml_record_backtrace (which is the
primitive), but this symbol is not declared publicly in the headers. All
occurrences in C, and the use-case detailed in the manual, are for
turning backtraces on, so introduce an actual C function to do this and
remove the declaration of the primitive.

dra27 commented:
Oops - typo corrections pushed. I did for `caml_sys_exit`, yes - although it should still just about work if something did, I think. The declaration of the primitive is inside a `CAML_INTERNALS` block, so the couple of cases which did use it actually declare it and I think a implicit declaration would just about work too, or perhaps only on 32bit?

dra27 commented:
Rebased to deal with #9948 (lots of things gone!) and #2195 (`caml_record_backtraces` takes an `int` instead of `void`).

dra27 commented:
This was left-over from work before 4.12. It's purpose was to tidy up the C API slightly so that when working in C you don't construct OCaml values and so forth to communicate with C functions.

I think it's stalled because I made a small API change post review.

This either wants to go in 4.13 or be dropped.

xavierleroy reviewed (approved):
Thanks for the update.  Looks good to me.  It could make sense to update `<caml/compatibility.h>`, removing `sys_exit` and `md5_string` and `md5_chan`.

dra27 commented:
Good idea - commits amended with those removals. Good to go with CI, thanks!
